[scroll-animations] ensure the `scroll-timeline` shorthand serializes according to the rule of a coordinating list property group

graouts opened the pull request:
#### d142435f83e61bec4b4d59894d6388902dd511ab
<pre>
[scroll-animations] ensure the `scroll-timeline` shorthand serializes according to the rule of a coordinating list property group
<a href="https://bugs.webkit.org/show_bug.cgi?id=265350">https://bugs.webkit.org/show_bug.cgi?id=265350</a>

Reviewed by Dean Jackson.

The `scroll-timeline` shorthand property must adhere to the rules of a coordinating list property
group (see <a href="https://drafts.csswg.org/css-values-4/#linked-properties).">https://drafts.csswg.org/css-values-4/#linked-properties).</a> The WPT test for serialization
of this property is incorrect (see <a href="https://github.com/web-platform-tests/wpt/issues/43336)">https://github.com/web-platform-tests/wpt/issues/43336)</a> so we
must fix it and ensure our implementation is right.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-shorthand.html:
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serialize):
(WebCore::ShorthandSerializer::serializeCoordinatingListPropertyGroup const): Introduce a new function
used to serialize a coordinating list property group.
(WebCore::ShorthandSerializer::serializeLayered const): Remove the logic specific to serialization of
`scroll-timeline-name` which no longer applies since this method won&apos;t be called for this property.

Canonical link: <a href="https://commits.webkit.org/271190@main">https://commits.webkit.org/271190@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5fc227a9c70a2275a3c1e699b77f9ba419657d83

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27661 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6299 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28907 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29881 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/25284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28135 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8238 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3692 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/25045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27927 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/5057 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/23730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/4386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/4559 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/24737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30521 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/25258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/25164 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/30675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/4577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/2722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/6074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/24471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3571 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/5005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

grorg commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -573,6 +575,45 @@ class LayerValues {
+        for (unsigned longhandIndex = 0; longhandIndex < length(); ++longhandIndex) {
+            auto& value = longhandValue(longhandIndex);
+            if (auto* valueList = dynamicDowncast<CSSValueList>(value)) {
+                auto* value = [&]() -> const CSSValue* {
```
It's a bit confusing that you reuse `value` here.

graouts replied:
I'll rename the second one when landing.

graouts committed:
[scroll-animations] ensure the `scroll-timeline` shorthand serializes according to the rule of a coordinating list property group
https://bugs.webkit.org/show_bug.cgi?id=265350

Reviewed by Dean Jackson.

The `scroll-timeline` shorthand property must adhere to the rules of a coordinating list property
group (see https://drafts.csswg.org/css-values-4/#linked-properties). The WPT test for serialization
of this property is incorrect (see https://github.com/web-platform-tests/wpt/issues/43336) so we
must fix it and ensure our implementation is right.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-shorthand.html:
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serialize):
(WebCore::ShorthandSerializer::serializeCoordinatingListPropertyGroup const): Introduce a new function
used to serialize a coordinating list property group.
(WebCore::ShorthandSerializer::serializeLayered const): Remove the logic specific to serialization of
`scroll-timeline-name` which no longer applies since this method won't be called for this property.

Canonical link: https://commits.webkit.org/271190@main
